File: src/quantum/benchmarks/grover_classical_comparison.py
# Quantum vs Classical Search Benchmark
#
# Compares Grover's algorithm (simulated) against classical search
# to verify the theoretical O(sqrt(N/M)) speedup.
#
# Reference: Grover, L. K. (1996). arXiv:quant-ph/9605043

from __future__ import annotations

import math
import random
import time
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable

# Try to import Rust bindings
try:
    import gororoba_py

    RUST_AVAILABLE = True
except ImportError:
    RUST_AVAILABLE = False


@dataclass
class SearchResult:
    """Result of a search benchmark."""

    method: str
    n_states: int
    n_marked: int
    oracle_calls: int
    wall_time_ms: float
    found_solution: bool
    theoretical_calls: float


@dataclass
class BenchmarkSummary:
    """Summary comparing quantum and classical search."""

    n_states: int
    n_marked: int
    classical_calls: int
    quantum_calls: int
    speedup_factor: float
    theoretical_speedup: float
    speedup_ratio: float  # actual / theoretical


def classical_random_search(
    oracle: Callable[[int], bool],
    n_states: int,
    n_marked: int,
    seed: int = 42,
) -> SearchResult:
    """Classical random search until a marked state is found.

    Args:
        oracle: Function returning True for marked states
        n_states: Total number of states
        n_marked: Number of marked states (for reference)
        seed: Random seed for reproducibility

    Returns:
        SearchResult with performance metrics
    """
    rng = random.Random(seed)
    calls = 0
    found = False
    start = time.perf_counter()

    # Random search with replacement
    visited = set()
    while len(visited) < n_states:
        idx = rng.randint(0, n_states - 1)
        if idx not in visited:
            visited.add(idx)
            calls += 1
            if oracle(idx):
                found = True
                break

    elapsed_ms = (time.perf_counter() - start) * 1000

    # Theoretical: E[calls] = N/M for random search
    theoretical = n_states / n_marked if n_marked > 0 else n_states

    return SearchResult(
        method="classical_random",
        n_states=n_states,
        n_marked=n_marked,
        oracle_calls=calls,
        wall_time_ms=elapsed_ms,
        found_solution=found,
        theoretical_calls=theoretical,
    )


def quantum_grover_search(
    marked_indices: list[int],
    n_states: int,
) -> SearchResult:
    """Quantum Grover search (simulated via Rust).

    Args:
        marked_indices: Indices of marked states
        n_states: Total number of states

    Returns:
        SearchResult with performance metrics
    """
    if not RUST_AVAILABLE:
        msg = "gororoba_py not available"
        raise ImportError(msg)

    n_qubits = int(math.ceil(math.log2(n_states)))
    n_marked = len(marked_indices)

    start = time.perf_counter()
    iterations, success_prob, candidates = gororoba_py.py_grover_search(
        n_qubits, marked_indices
    )
    elapsed_ms = (time.perf_counter() - start) * 1000

    # Check if any candidate is marked
    found = any(c in marked_indices for c in candidates)

    # Theoretical: O(sqrt(N/M)) iterations
    theoretical = (math.pi / 4) * math.sqrt(n_states / n_marked) if n_marked > 0 else 1

    return SearchResult(
        method="quantum_grover",
        n_states=n_states,
        n_marked=n_marked,
        oracle_calls=iterations,
        wall_time_ms=elapsed_ms,
        found_solution=found,
        theoretical_calls=theoretical,
    )


def compare_search_methods(
    n_states: int,
    n_marked: int,
    seed: int = 42,
) -> BenchmarkSummary:
    """Compare quantum and classical search on the same problem.

    Args:
        n_states: Total search space size (should be power of 2)
        n_marked: Number of marked states
        seed: Random seed

    Returns:
        BenchmarkSummary comparing the methods
    """
    # Ensure power of 2
    n_qubits = int(math.ceil(math.log2(n_states)))
    n_states = 2**n_qubits

    # Generate random marked states
    rng = random.Random(seed)
    marked_indices = rng.sample(range(n_states), n_marked)
    marked_set = set(marked_indices)

    def oracle(idx: int) -> bool:
        return idx in marked_set

    # Run classical search
    classical_result = classical_random_search(oracle, n_states, n_marked, seed)

    # Run quantum search
    if RUST_AVAILABLE:
        quantum_result = quantum_grover_search(marked_indices, n_states)
    else:
        # Fallback: estimate based on theoretical formula
        theoretical_iters = int(
            round((math.pi / 4) * math.sqrt(n_states / n_marked))
        )
        quantum_result = SearchResult(
            method="quantum_grover_theoretical",
            n_states=n_states,
            n_marked=n_marked,
            oracle_calls=theoretical_iters,
            wall_time_ms=0.0,
            found_solution=True,
            theoretical_calls=theoretical_iters,
        )

    # Compute speedup
    speedup = classical_result.oracle_calls / quantum_result.oracle_calls
    theoretical_speedup = math.sqrt(n_states / n_marked)
    speedup_ratio = speedup / theoretical_speedup

    return BenchmarkSummary(
        n_states=n_states,
        n_marked=n_marked,
        classical_calls=classical_result.oracle_calls,
        quantum_calls=quantum_result.oracle_calls,
        speedup_factor=speedup,
        theoretical_speedup=theoretical_speedup,
        speedup_ratio=speedup_ratio,
    )


def run_benchmark_suite(
    sizes: list[int] | None = None,
    marked_fractions: list[float] | None = None,
) -> list[BenchmarkSummary]:
    """Run benchmark suite across multiple problem sizes.

    Args:
        sizes: List of search space sizes (powers of 2)
        marked_fractions: Fractions of states to mark

    Returns:
        List of benchmark summaries
    """
    if sizes is None:
        sizes = [64, 256, 1024, 4096]
    if marked_fractions is None:
        marked_fractions = [0.01, 0.05, 0.1, 0.25]

    results = []
    for n_states in sizes:
        for frac in marked_fractions:
            n_marked = max(1, int(n_states * frac))
            summary = compare_search_methods(n_states, n_marked)
            results.append(summary)

    return results


def format_benchmark_table(results: list[BenchmarkSummary]) -> str:
    """Format benchmark results as a table.

    Args:
        results: List of benchmark summaries

    Returns:
        Formatted table string
    """
    lines = [
        "| N_states | N_marked | Classical | Quantum | Speedup | Theory | Ratio |",
        "|----------|----------|-----------|---------|---------|--------|-------|",
    ]

    for r in results:
        lines.append(
            f"| {r.n_states:8d} | {r.n_marked:8d} | {r.classical_calls:9d} | "
            f"{r.quantum_calls:7d} | {r.speedup_factor:7.2f} | "
            f"{r.theoretical_speedup:6.2f} | {r.speedup_ratio:5.2f} |"
        )

    return "\n".join(lines)


if __name__ == "__main__":
    print("Quantum vs Classical Search Benchmark")
    print("=" * 60)
    print()

    if not RUST_AVAILABLE:
        print("WARNING: gororoba_py not available, using theoretical estimates")
        print()

    results = run_benchmark_suite()
    print(format_benchmark_table(results))
    print()
    print("Speedup ratio near 1.0 indicates algorithm matches theory.")

File: crates/gororoba_cli_quantum/src/bin/grover-classical-comparison.rs
use anyhow::Result;
use clap::Parser;
use quantum_core::grover::{grover_search, GroverConfig, optimal_iterations};
use rand::prelude::*;
use std::collections::HashSet;

#[derive(Parser, Debug)]
#[command(author, version, about = "Quantum vs Classical Search Benchmark")]
struct Args {
    #[arg(short, long, default_value = "64,256,1024,4096")]
    sizes: String,

    #[arg(short, long, default_value = "0.01,0.05,0.1,0.25")]
    fractions: String,
    
    #[arg(long, default_value_t = 42)]
    seed: u64,
}

struct BenchmarkSummary {
    n_states: usize,
    n_marked: usize,
    classical_calls: usize,
    quantum_calls: usize,
    speedup_factor: f64,
    theoretical_speedup: f64,
    speedup_ratio: f64,
}

fn classical_random_search(n_states: usize, marked_set: &HashSet<usize>, seed: u64) -> usize {
    let mut rng = StdRng::seed_from_u64(seed);
    let mut calls = 0;
    let mut visited = HashSet::new();

    while visited.len() < n_states {
        let idx = rng.gen_range(0..n_states);
        if !visited.contains(&idx) {
            visited.insert(idx);
            calls += 1;
            if marked_set.contains(&idx) {
                break;
            }
        }
    }
    calls
}

fn run_benchmark_suite(sizes: &[usize], fractions: &[f64], seed: u64) -> Vec<BenchmarkSummary> {
    let mut results = Vec::new();

    for &n_states in sizes {
        let n_qubits = (n_states as f64).log2().ceil() as u32;
        let actual_n_states = 1usize << n_qubits;
        
        let mut rng = StdRng::seed_from_u64(seed);

        for &frac in fractions {
            let n_marked = 1.max((actual_n_states as f64 * frac) as usize);

            let mut indices: Vec<usize> = (0..actual_n_states).collect();
            indices.shuffle(&mut rng);
            let marked_indices: Vec<usize> = indices.into_iter().take(n_marked).collect();
            let marked_set: HashSet<usize> = marked_indices.iter().copied().collect();

            let classical_calls = classical_random_search(actual_n_states, &marked_set, seed);

            let q_iters = optimal_iterations(actual_n_states, n_marked);
            let is_marked = |x: usize| marked_set.contains(&x);
            let q_result = grover_search(
                n_qubits as usize,
                is_marked,
                n_marked,
                GroverConfig { iterations: Some(q_iters), top_k: 1 }
            );

            let quantum_calls = q_result.iterations;
            let theoretical_speedup = (actual_n_states as f64 / n_marked as f64).sqrt();
            let speedup = classical_calls as f64 / quantum_calls.max(1) as f64;
            let speedup_ratio = speedup / theoretical_speedup;

            results.push(BenchmarkSummary {
                n_states: actual_n_states,
                n_marked,
                classical_calls,
                quantum_calls,
                speedup_factor: speedup,
                theoretical_speedup,
                speedup_ratio,
            });
        }
    }
    results
}

fn main() -> Result<()> {
    let args = Args::parse();
    
    let sizes: Vec<usize> = args.sizes.split(',')
        .map(|s| s.trim().parse::<usize>().expect("Invalid size"))
        .collect();
        
    let fractions: Vec<f64> = args.fractions.split(',')
        .map(|s| s.trim().parse::<f64>().expect("Invalid fraction"))
        .collect();

    println!("Quantum vs Classical Search Benchmark");
    println!("============================================================");
    println!();

    let results = run_benchmark_suite(&sizes, &fractions, args.seed);

    println!("| N_states | N_marked | Classical | Quantum | Speedup | Theory | Ratio |");
    println!("|----------|----------|-----------|---------|---------|--------|-------|");

    for r in results {
        println!(
            "| {:8} | {:8} | {:9} | {:7} | {:7.2} | {:6.2} | {:5.2} |",
            r.n_states,
            r.n_marked,
            r.classical_calls,
            r.quantum_calls,
            r.speedup_factor,
            r.theoretical_speedup,
            r.speedup_ratio
        );
    }

    println!();
    println!("Speedup ratio near 1.0 indicates algorithm matches theory.");

    Ok(())
}
